pymchelper/averaging.py

```diff
@@ -0,0 +1,214 @@
+Moreover, in case averaging is employed, we can estimate spread of the data as standard deviation or standard error.
+
+The binary output files from each job may be quite large (even ~GB for scoring in fine 3D mesh), to obtain good
+statistics we sometimes parallelise the simulation of hundreds or thousands of jobs (when using HPC clusters).
+In such case it's not feasible to load all the data into memory and calculate the average in one go, using standard
+functions from numpy library. Instead, we need to calculate the average in an online manner,
+i.e. by updating the state of respective aggregator object with each new binary output file read.
+
+Such approach results in a significant reduction of memory usage and is more numerically stable.
+
+This module contains several classes for aggregating data from multiple files:
+- Aggregator: base class for all other aggregators
+- WeightedStatsAggregator: for calculating weighted (using reliability, not frequency weights) mean and variance
+- ConcatenatingAggregator: for concatenating data
+- SumAggregator: for calculating sum instead of variance
+- NoAggregator: for cases when no aggregation is required
+
+All aggregators have `data` and `error` property, which can be used to obtain the result of the aggregation.
+The `data` property returns the result of the aggregation: mean, sum or concatenated array.
+The `error` property returns the spread of data for WeightedStatsAggregator, and `None` for other aggregators.
+
+The `update` method is used to update the state of the aggregator with new data from the file.
+"""
+
+from dataclasses import dataclass, field
+import logging
+from typing import Union
+import numpy as np
+from numpy.typing import ArrayLike
+
+
+@dataclass
+class Aggregator:
+    """
+    Base class for all aggregators.
+    The `data` property returns the result of the aggregation, needs to be implemented in derived classes.
+    The `error` function returns the spread of data, can be implemented in derived classes. It's a function,
+    not a property as different type of error can be calculated (standard deviation, standard error, etc.).
+    Type of errors may be then passed in optional keyword arguments `**kwargs`.
+    """
+
+    data: Union[float, ArrayLike] = float('nan')
+    _updated: bool = field(default=False, repr=False, init=False)
+
+    def update(self, value: Union[float, ArrayLike], **kwargs):
+        """Update the state of the aggregator with new data."""
+        raise NotImplementedError(f"Update function not implemented for {self.__class__.__name__}")
+
+    def error(self, **kwargs):
+        """Default implementation of error function, returns None."""
+        logging.debug("Error calculation not implemented for %s", self.__class__.__name__)
+
+    @property
+    def updated(self):
+        """Check if the aggregator was updated."""
+        if isinstance(self.data, float) and np.isnan(self.data):
```

Comment:
is NANs here used to trigger update? (maybe describe the logic in docstring)

Reply:
NaN is initial value assigned to data when constructing dataclass object, see:
https://github.com/DataMedSci/pymchelper/blob/93cfda176bce289504c2a72bb1122166f6e080ac/pymchelper/averaging.py#L48

when the object is constructed and update method is not called yet we are in the not-updated-yet state.

anyway good point - I will add more comments